### PR TITLE
En 5751/10% of fees to leader

### PR DESCRIPTION
**Pull request by sasurobert**
10% of fees to leader.
All accumulated fees kept in metablock through all the epoch.

### CODE REVIEW

File: process/block/metablock.go [hunk 1]
```diff
@@ -1506,8 +1520,44 @@ func (mp *metaProcessor) createPeerInfo() ([]block.PeerData, error) {
...
 }
 
+func (mp *metaProcessor) verifyTotalAccumulatedFeesInEpoch(blockChain data.ChainHandler, metaHdr *block.MetaBlock) error {
+	computedTotalFees, err := mp.computeAccumulatedFeesInEpoch(blockChain, metaHdr)
+	if err != nil {
+		return err
+	}
+
+	if computedTotalFees.Cmp(metaHdr.AccumulatedFeesInEpoch) != 0 {
+		return process.ErrAccumulatedFeesDoNotMatch
```
Comment (iulianpascalau) [inline process/block/metablock.go]: Might be a good idea to wrap here the error, something like:
```GO
return fmt.Errorf("%w, got %v, computed %v", process.ErrAccumulatedFeesDoNotMatch, metaHdr.AccumulatedFeesInEpoch, computedTotalFees)
```

File: process/block/metablock.go [hunk 2]
```diff
@@ -1506,8 +1520,44 @@ func (mp *metaProcessor) createPeerInfo() ([]block.PeerData, error) {
...
+		if !ok {
+			return nil, process.ErrWrongTypeAssertion
+		}
+
+		if !lastHdr.IsStartOfEpochBlock() {
+			currentlyAccumulated = lastMeta.AccumulatedFeesInEpoch
+		}
+	}
+
+	currentlyAccumulated.Add(currentlyAccumulated, metaHdr.GetAccumulatedFees())
```
Comment (iulianpascalau) [inline process/block/metablock.go]: efficient way :+1:

File: process/interface.go [hunk 3]
```diff
@@ -203,7 +203,7 @@ type BlockProcessor interface {
 	CreateNewHeader() data.HeaderHandler
 	CreateBlockBody(initialHdrData data.HeaderHandler, haveTime func() bool) (data.BodyHandler, error)
 	RestoreBlockIntoPools(header data.HeaderHandler, body data.BodyHandler) error
-	ApplyBodyToHeader(hdr data.HeaderHandler, body data.BodyHandler) (data.BodyHandler, error)
+	ApplyBodyToHeader(blockChain data.ChainHandler, hdr data.HeaderHandler, body data.BodyHandler) (data.BodyHandler, error)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/interface.go]: done.

File: cmd/node/config/economics.toml [hunk 4]
```diff
@@ -4,10 +4,7 @@
     BurnAddress = "deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef"
 
 [RewardsSettings]
-    RewardsValue = "20000000000000000000" #20ERD
-    CommunityPercentage = 0.10
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/config/economics.toml]: wow, good question. asked to see how to handle.

[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/config/economics.toml]: not more needed.

File: data/state/interface.go [hunk 5]
```diff
@@ -69,6 +71,7 @@ type AccountHandler interface {
 //  with some extra features like signing statistics or rating information
 type PeerAccountHandler interface {
 	AccountHandler
+	IncreaseAccumulatedFees(value *big.Int) error
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/state/interface.go]: here is really increase.

[the remark this answers is not in the document]
Comment (sasurobert) [reply data/state/interface.go]: verified before to be increase.

File: core/computers.go [hunk 6]
```diff
@@ -55,3 +57,16 @@ func MinUint64(a uint64, b uint64) uint64 {
 	}
 	return b
 }
+
+// GetPercentageOfValue returns the percentage part of the value
+func GetPercentageOfValue(value *big.Int, percentage float64) *big.Int {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/computers.go]: as we have in a single function, we can easily refactor when changing in all protocol.

File: process/block/metablock.go (hunk 1 above)
Comment (sasurobert) [reply process/block/metablock.go]: done.

File: process/block/metablock.go [hunk 7]
```diff
@@ -1506,6 +1518,42 @@ func (mp *metaProcessor) createPeerInfo() ([]block.PeerData, error) {
...
+
+	if computedTotalFees.Cmp(metaHdr.AccumulatedFeesInEpoch) != 0 {
+		return process.ErrAccumulatedFeesDoNotMatch
+	}
+
+	return nil
+}
+
+func (mp *metaProcessor) computeAccumulatedFeesInEpoch(metaHdr *block.MetaBlock) (*big.Int, error) {
+	currentlyAccumulated := big.NewInt(0)
```
Comment (SebastianMarian) [inline process/block/metablock.go]: currentlyAccumulatedFees or currentlyAccumulatedFeesInEpoch

File: process/block/metablock.go [hunk 8]
```diff
@@ -1521,6 +1569,13 @@ func (mp *metaProcessor) ApplyBodyToHeader(hdr data.HeaderHandler, bodyHandler d
 		return nil, process.ErrWrongTypeAssertion
 	}
 
+	if check.IfNil(bodyHandler) {
+		return nil, process.ErrNilBlockBody
+	}
+	if check.IfNil(mp.blockChain) {
```
Comment (SebastianMarian) [inline process/block/metablock.go]: This check could be removed as it is checked for nil on constructor

File: process/block/metablock.go [hunk 9]
```diff
@@ -1521,6 +1569,13 @@ func (mp *metaProcessor) ApplyBodyToHeader(hdr data.HeaderHandler, bodyHandler d
 		return nil, process.ErrWrongTypeAssertion
 	}
 
+	if check.IfNil(bodyHandler) {
```
Comment (SebastianMarian) [inline process/block/metablock.go]: You can move this check in the first line and also add a similar check for headerHandler

File: data/state/interface.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/state/interface.go]: done.

File: integrationTests/testInitializer.go [hunk 10]
```diff
@@ -249,6 +249,8 @@ func CreateMetaChain() data.ChainHandler {
 		badBlockCache,
 	)
 	metaChain.GenesisBlock = &dataBlock.MetaBlock{}
+	genesisHeaderM, _ := TestMarshalizer.Marshal(metaChain.GenesisBlock)
```
Comment (SebastianMarian) [inline integrationTests/testInitializer.go]: genesisHeaderHash, _ := core.CalculateHash(TestMarshalizer, TestHasher, metaChain.GenesisBlock)
	metaChain.SetGenesisHeaderHash(genesisHeaderHash)

File: data/state/peerAccount.go [hunk 11]
```diff
@@ -335,6 +337,34 @@ func (pa *PeerAccount) DecreaseValidatorSuccessRateWithJournal(value uint32) err
 	return pa.accountTracker.SaveAccount(pa)
 }
 
+// IncreaseLeaderAccumulatedFees increases the account's accumulated fees
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/state/peerAccount.go]: done.

File: core/computers.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply core/computers.go]: ok

File: data/state/peerJournalEntries.go [hunk 12]
```diff
@@ -411,3 +411,35 @@ func (pjec *PeerJournalEntryUnStakedNonce) Revert() (AccountHandler, error) {
 func (pjec *PeerJournalEntryUnStakedNonce) IsInterfaceNil() bool {
 	return pjec == nil
 }
+
+//------- PeerJournalEntryStake
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/state/peerJournalEntries.go]: done.

File: data/state/peerJournalEntries.go [hunk 13]
```diff
@@ -411,3 +411,35 @@ func (pjec *PeerJournalEntryUnStakedNonce) Revert() (AccountHandler, error) {
...
+
+// PeerJournalEntryAccumulatedFee is used to revert a stake change
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/state/peerJournalEntries.go]: done.

File: data/state/peerJournalEntries.go [hunk 14]
```diff
@@ -411,3 +411,35 @@ func (pjec *PeerJournalEntryUnStakedNonce) Revert() (AccountHandler, error) {
...
+	}
+
+	return &PeerJournalEntryAccumulatedFee{
+		account:            account,
+		oldAccumulatedFees: oldAccumulatedFees,
+	}, nil
+}
+
+// Revert applies undo operation
+func (pjes *PeerJournalEntryAccumulatedFee) Revert() (AccountHandler, error) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/state/peerJournalEntries.go]: done.

File: data/state/peerJournalEntries.go [hunk 15]
```diff
@@ -411,3 +411,35 @@ func (pjec *PeerJournalEntryUnStakedNonce) Revert() (AccountHandler, error) {
...
+	pjes.account.AccumulatedFees = pjes.oldAccumulatedFees
+
+	return pjes.account, nil
+}
+
+// IsInterfaceNil returns true if there is no value under the interface
+func (pjes *PeerJournalEntryAccumulatedFee) IsInterfaceNil() bool {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/state/peerJournalEntries.go]: done.

File: process/block/metablock.go (hunk 7 above)
Comment (sasurobert) [reply process/block/metablock.go]: done.

File: process/block/metablock.go (hunk 8 above)
Comment (sasurobert) [reply process/block/metablock.go]: done.

File: process/block/metablock.go (hunk 9 above)
Comment (sasurobert) [reply process/block/metablock.go]: done.

File: integrationTests/testInitializer.go (hunk 10 above)
Comment (sasurobert) [reply integrationTests/testInitializer.go]: done.

**Pull request merged**